File: urjc/cp4-f.cpp
#include <iostream>
#include <cstring>
#include <map>
#include <vector>

using namespace std;

vector<int> contagiar(vector<int> contagios, map<int, vector<int>> contactos, int idx) {
    for (vector<int>::iterator it = contactos[idx].begin(); it != contactos[idx].end(); it++) {
        if (contagios[*it] == 0) {
            contagios[*it] = 1;
            contagios = contagiar(contagios, contactos, *it);
        }
    }

    return contagios;
}

int main() {
    int n, m, a, b;
    bool flag;
    
    map<int, vector<int>> contactos;
    char frase[9];

    cin >> n >> m;

    vector<int> contagios(n, 0);

    for(int i=0; i<m; i++) {
        cin >> frase;

        if (strcmp(frase, "CONTACTO") == 0) {
            cin >> a >> b;

            contactos[a].push_back(b);
            contactos[b].push_back(a);
            if (contagios[a] == 1) {
                
                if(contagios[b] == 0) {
                    contagios[b] = 1;
                    contagios = contagiar(contagios, contactos, b);
                }
            }
            else if(contagios[b] == 1) {
                contagios[a] = 1;
                contagios = contagiar(contagios, contactos, a);
            }
        }
        else if(strcmp(frase, "POSITIVO") == 0) {
            cin >> a;
            contagios[a] = 1;
            contagios = contagiar(contagios, contactos, a);
        }

        else {
            cin >> a;
            if (contagios[a] == 1) {
                cout << "POSIBLE" << endl;
            }
            else {
                cout << "NO" << endl;
            }
        }

    }

    return 0;
}

File: urjc/cp4-f.py
def contagiar(contagios, contactos, n):
    for ele in contactos[n]:
        if contagios[ele] == 0:
            contagios[ele] = 1
            contagiar(contagios, contactos, ele)


n, m = list(map(lambda x: int(x), input().split()))
contagios = [0]*n
contactos = {}

if n > 0:
    for i in range(0, m):
        l = input().split()
        if l[0] == 'CONTACTO':
            a = int(l[1])
            b = int(l[2])
            if a not in contactos:
                contactos[a] = []

            contactos[a].append(b)

            if b not in contactos:
                contactos[b] = []

            contactos[b].append(a)

            if contagios[a] == 1:
                if contagios[b] == 0:
                    contagios[b] = 1
                    contagiar(contagios, contactos, b)

            elif contagios[b] == 1:
                contagios[a] = 1
                contagiar(contagios, contactos, a)

        elif l[0] == 'POSITIVO':
            a = int(l[1])
            contagios[a] = 1
            contagiar(contagios, contactos, a)

        else:
            a = int(l[1])
            if contagios[a] == 1:
                print("POSIBLE")
            else:
                print("NO")